File: 2115. Find All Possible Recipes from Given Supplies/main.py
from typing import List


class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        recipes_map = {}
        for i in range(len(recipes)):
            recipes_map[recipes[i]] = ingredients[i]
        supplies_set = set(supplies)

        visit = set()

        def can_create_recipe(recipe):
            if recipe in supplies_set:
                return True
            for ingr in recipes_map[recipe]:
                if ingr in visit:
                    return False
                if ingr in recipes_map:
                    visit.add(ingr)
                    can_create_recipe(ingr)
                    visit.remove(ingr)
                if ingr not in supplies_set:
                    return False
            supplies_set.add(recipe)
            return True
        
        res = []
        for r in recipes:
            if can_create_recipe(r):
                res.append(r)

        return res

File: 2115. Find All Possible Recipes from Given Supplies/main.go
package main

func findAllRecipes(recipes []string, ingredients [][]string, supplies []string) []string {
	canCook := make(map[string]bool)
	for _, s := range supplies {
		canCook[s] = true
	}
	recipeIndex := make(map[string]int)
	for i, r := range recipes {
		recipeIndex[r] = i
	}

	var dfs func(r string) bool
	dfs = func(r string) bool {
		if val, exists := canCook[r]; exists {
			return val
		}
		if _, exists := recipeIndex[r]; !exists {
			return false
		}

		canCook[r] = false // handle circle

		for _, ingr := range ingredients[recipeIndex[r]] {
			if !dfs(ingr) {
				return false
			}
		}

		canCook[r] = true
		return true
	}

	res := make([]string, 0)

	for _, r := range recipes {
		if dfs(r) {
			res = append(res, r)
		}
	}

	return res
}
